libBlueSky/SizeGrip.cpp

```diff
@@ -0,0 +1,224 @@
+#include "SizeGrip.hpp"
+
+#include <QAbstractScrollArea>
+#include <QApplication>
+#include <QDesktopWidget>
+#include <QLayout>
+#include <QMouseEvent>
+
+namespace BlueSky {
+
+SizeGrip::SizeGrip(QWidget* parent)
+    : QSizeGrip(parent)
+    , mResizeDirections(ResizeDefault)
+    , mParentWindow(parent)
+    , mGotMousePress(false)
+    , dxMax(0)
+    , dyMax(0)
+    , mCorner(mParentWindow->isLeftToRight() ? Qt::BottomRightCorner : Qt::BottomLeftCorner)
+{
+    Q_ASSERT(parent);
+    resize(sizeHint());
+}
+
+SizeGrip::ResizeDirections SizeGrip::resizeBehaviour() const
+{
+    return mResizeDirections;
+}
+
+void SizeGrip::setResizeBehaviour(int behaviour)
+{
+    mResizeDirections = static_cast<ResizeDirections>(behaviour);
+}
+
+void SizeGrip::mousePressEvent(QMouseEvent* ev)
+{
+    if (ev->button() != Qt::LeftButton) {
+        QWidget::mousePressEvent(ev);
+        return;
+    }
+
+    QWidget *tlw = mParentWindow;
+    mGotMousePress = true;
+    mStartPos = ev->globalPos();
+    mDlgRect = mParentWindow->geometry();
+
+#ifdef Q_WS_X11
+    // Use a native X11 sizegrip for "real" top-level windows if supported.
+    if (tlw->isWindow() && X11->isSupportedByWM(ATOM(_NET_WM_MOVERESIZE))
+        && !(tlw->windowFlags() & Qt::X11BypassWindowManagerHint)
+        && !tlw->testAttribute(Qt::WA_DontShowOnScreen) && !qt_widget_private(tlw)->hasHeightForWidth()) {
+        XEvent xev;
+        xev.xclient.type = ClientMessage;
+        xev.xclient.message_type = ATOM(_NET_WM_MOVERESIZE);
+        xev.xclient.display = X11->display;
+        xev.xclient.window = tlw->winId();
+        xev.xclient.format = 32;
+        xev.xclient.data.l[0] = ev->globalPos().x();
+        xev.xclient.data.l[1] = ev->globalPos().y();
+        if (atBottom())
+            xev.xclient.data.l[2] = atLeft() ? 6 : 4; // bottomleft/bottomright
+        else
+            xev.xclient.data.l[2] = atLeft() ? 0 : 2; // topleft/topright
+        xev.xclient.data.l[3] = Button1;
+        xev.xclient.data.l[4] = 0;
+        XUngrabPointer(X11->display, X11->time);
+        XSendEvent(X11->display, QX11Info::appRootWindow(x11Info().screen()), False,
+                   SubstructureRedirectMask | SubstructureNotifyMask, &xev);
+        return;
+    }
+#endif
+#ifdef Q_WS_WIN
+    if (tlw->isWindow() && !tlw->testAttribute(Qt::WA_DontShowOnScreen) && !qt_widget_private(tlw)->hasHeightForWidth()) {
```

Comment:
Yeah, that's another good argument. Hm ... if the following works, I can combine it with Qt's "original" implementation (like it should be) and write something like this:

```c++
void SizeGrip::mouseMoveEvent(QMouseEvent* ev) {
  QSizeGrip::mouseMoveEvent(ev);
  // this calls tlw->setGeometry(), which is the whole reason for the crappy implementation above. But ... if I do the calculation again, I can tweak it.
  
  QRect nr = tlw->rect(); // <--- should work

  // recalc delta-x and delta-y to move/resize top-left according to the behavioural flags

  // finally call tlw->setGeometry() again ...
  tlw->setGeometry(nr);
```

This would be more harmless. Anyway, I opened a feature request to Qt - my first ever :smile:. You might want to have a look [here](https://bugreports.qt.io/browse/QTBUG-47443).